Add KubeVirt cloud-init network data support

qinqon opened the pull request:
Add support at kubevirt for parsing [cloud-init both config drive network data](https://cloudinit.readthedocs.io/en/latest/reference/network-config.html#network-configuration-sources) wich
follows openstack network JSON [format](https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json
) and also no cloud network data this is translated to
dracut kernel arguments: physical ethernet interfaces, static IP
configuration, DHCP, gateways, and nameservers.

Also expose first interface IPv4 and IPv6 addresses as KUBEVIRT_IPV4 and
KUBEVIRT_IPV6 attributes in the metadata provider. This allows users
to access the assigned IP addresses from the cloud-init network

Additionally, includes a dracut module enhancement that preloads the virtio_blk module in the afterburn-network-kargs service to ensure proper disk access during early boot in virtualized environments, for rhcos this module is not mounted before the systemd unit runs.

**Note for reviewers:**
  This PR includes a refactor of the KubeVirt provider to follow the ProxmoxVE provider structure, which served as inspiration for implementing network data support. The changes organize the code into
  separate modules and allow to test the whole implementation passing a test fixture that contains meta_data.json and network_data.json similar to how proxmoxve is testing it.


**Supported cloud-init network data subset:**
  - Static IPv4/IPv6 address configuration with subnets
  - Default gateway configuration (IPv4/IPv6)
  - Physical ethernet interface configuration
  - DHCP configuration
  - DNS configuration
 
 **Testing:**
  - Added IPv6 test cases
  - Validates network configuration parsing and application

qinqon committed:
dracut: Add virtio_blk module preload to afterburn-network-kargs service

Add ExecStartPre=modprobe virtio_blk to the afterburn-network-kargs.service
to ensure the virtio_blk kernel module is loaded before the service starts.
This is required for the kubevirt provider to successfully mount the config
drive disk device.

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon committed:
kubevirt: Refactor the provider to follow the proxmoxve structure

The proxmoxve provider has a well-structured design where the provider's
public interface is a ConfigDrive that handles mounting the config drive
device, while the cloud-init parsing logic is isolated in cloudinit.rs.
This parser expects a path to a cloud-init directory, which makes testing
the trait interfaces directly much easier.

This change prepares the kubevirt provider to add support for cloud-init
network data that will be parsed similarly to proxmoxve.

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon committed:
kubevirt: Support config drive network data

Add support at kubevirt for parsing cloud-init config drive network data wich
follows openstack network JSON format [1][2] this is translated to
dracut kernel arguments: physical ethernet interfaces, static IP
configuration, DHCP, gateways, and nameservers.

Also expose first interface IPv4 and IPv6 addresses as KUBEVIRT_IPV4 and
KUBEVIRT_IPV6 attributes in the metadata provider. This allows users
to access the assigned IP addresses from the cloud-init network

[1] https://cloudinit.readthedocs.io/en/latest/reference/network-config.html#network-configuration-sources
[2] https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon committed:
docs: Add release notes

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon commented on src/providers/kubevirt/cloudconfig.rs:
```diff
@@ -0,0 +1,317 @@
+    fn ssh_keys(&self) -> Result<Vec<PublicKey>> {
+        let default_map = HashMap::new();
+        let public_keys_map = self.meta_data.public_keys.as_ref().unwrap_or(&default_map);
+        let public_keys_vec: Vec<&String> = public_keys_map.values().collect();
+        let mut out = vec![];
+        for key in public_keys_vec {
+            let key = PublicKey::parse(key)?;
+            out.push(key);
+        }
+        Ok(out)
+    }
```
@gemini-code-assist this compiles into an error
```
:!cargo build 2>&1| tee /tmp/nvim.ellorent/LqADgv/3                                                                                                                                                       
   Compiling afterburn v5.9.0 (/home/ellorent/Documents/cnv/upstream/afterburn)                                                                                                                                    
error[E0599]: the method `as_deref` exists for enum `Option<HashMap<String, String>>`, but its trait bounds were not satisfied                                                                                     
   --> src/providers/kubevirt/cloudconfig.rs:102:14                                                                                                                                                                
    |                                                                                                                                                                                                              
100 | /         self.meta_data                                                                                                                                                                                     
101 | |             .public_keys                                                                                                                                                                                   
102 | |             .as_deref()                                                                                                                                                                                    
    | |_____________-^^^^^^^^                                                                                                                                                                                      
    |                                                                                                                                                                                                              
   ::: /home/ellorent/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:230:1                                                                        
    |                                                                                                                                                                                                              
230 |   pub struct HashMap<K, V, S = RandomState> {                                                                                                                                                                
    |   ----------------------------------------- doesn't satisfy `_: Deref`                                                                                                                                       
    |                                                                                                                                                                                                              
    = note: the following trait bounds were not satisfied:                                                                                                                                                         
            `HashMap<std::string::String, std::string::String>: Deref`                                                                                                                                             
                                                                                                                                                                                                                   
For more information about this error, try `rustc --explain E0599`.                                                                                                                                                
error: could not compile `afterburn` (bin "afterburn") due to 1 previous error                                                                                                                                     
                                                                                                                                                                                                                   
(1 of 14) error 599: the method `as_deref` exists for enum `Option<HashMap<String, String>>`, but its trait bounds were not satisfied                                                                   
```

qinqon commented on src/providers/kubevirt/cloudconfig.rs:
```diff
@@ -0,0 +1,310 @@
+            attrs.insert("KUBEVIRT_INSTANCE_TYPE".to_string(), instance_type.clone());
+        }
+
+        if let Some(interface_with_ips) = self
+            .networks()?
+            .iter()
+            .find(|iface| !iface.ip_addresses.is_empty())
+        {
+            interface_with_ips
+                .ip_addresses
+                .iter()
+                .for_each(|ip| match ip {
+                    IpNetwork::V4(network) => {
+                        attrs
+                            .entry("KUBEVIRT_IPV4".to_owned())
+                            .or_insert_with(|| network.ip().to_string());
+                    }
+                    IpNetwork::V6(network) => {
+                        attrs
+                            .entry("KUBEVIRT_IPV6".to_owned())
+                            .or_insert_with(|| network.ip().to_string());
+                    }
+                });
+        }
```
This is the same on proxmoxve provider, we take the first interface to set this variables, on multi interface env with addresses on multiple interface, there is no clear view of which one we have to return, doing as proxmoxve is the most we can do since this bug exist there too.

qinqon commented on src/providers/kubevirt/networkdata.rs:
```diff
@@ -0,0 +1,360 @@
+                .map(|ip| IpAddr::from_str(ip))
+                .collect::<Result<Vec<IpAddr>, AddrParseError>>()?;
+        }
+    }
```
We are just using the "networks()" call to later on generate kernel args where we have just one global nameservers entries, the logic there just look for nameservers at all interfaces so this just works fine.

qinqon commented on src/providers/kubevirt/cloudconfig.rs:
```diff
@@ -0,0 +1,281 @@
+                iface_name
+            } else {
+                continue;
+            };
```
This is compiling allright

qinqon commented on src/providers/kubevirt/networkdata.rs:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    Ok(interfaces)
+}
```
The networkData in the context of fedoracores is kind of bootstrap networking, full blown networking should be done with ignition for example dumping stuff at /etc/nmstate

qinqon commented on src/providers/kubevirt/networkdata.rs:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    Ok(interfaces)
+}
```
This still would be kind of unexpected since users don not expect interface name to be used to sort things up, proper thing would be to use an indexmap that keeps the insertion order
but since this is expected to be a super minimal configuration I don't thing is worth it.

Something like
```rust
use serde::Deserialize;
use serde_yaml;
use indexmap::IndexMap;

#[derive(Debug, Deserialize)]
struct MyData {
    // This will preserve the insertion order of the keys
    #[serde(with = "indexmap::serde_seq")]
    pub items: IndexMap<String, String>,
}

fn main() {
    let yaml_string = r#"
    items:
        key_c: value_c
        key_a: value_a
        key_b: value_b
    "#;

    let data: MyData = serde_yaml::from_str(yaml_string).unwrap();

    // The keys will be iterated in the order they were defined in the YAML
    // "key_c", "key_a", "key_b"
    for (key, value) in data.items.iter() {
        println!("{}: {}", key, value);
    }
}
```

qinqon replied:
Anyhow I will add a extra commit moving to indexmap in case maintainers things is something worth it.

jlebon commented on dracut/30afterburn/afterburn-network-kargs.service:
```diff
@@ -14,6 +14,8 @@ OnFailureJobMode=isolate
 [Service]
 Environment=AFTERBURN_OPT_PROVIDER=--cmdline
+# To read config drive at kubevirt provider
+ExecStartPre=modprobe virtio_blk
```
Hmm, I don't think this should be done unconditionally on every platform. Couldn't Afterburn itself do this in the kubevirt provider path?

prestist replied:
Ugh Idk how I did not see your comment but yes, great catch. Ill open a followup.

qinqon replied:
Ahh right, we totally forgot about that

prestist reviewed (commented):
From my perspective, this LGTM, and thank you for contributing!

 Regarding cidata did you want to add it to this pr or a follow up pr? and forgive me but could you could you elaborate on the v1 , v2 support NoCloud? I guess what I am getting at is it common for v1 to be used still?

qinqon committed:
kubevirt: Add NoCloud network configuration support

Extend the KubeVirt provider to support NoCloud network configuration
format v1 and v2 both yaml and json [1].

The NoCloud looks for a device with label "cidata" and there at the
/root path for files `meta-data` and `network-configuration`.

[1] https://cloudinit.readthedocs.io/en/latest/reference/network-config.html#network-configuration-sources

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

prestist reviewed (approved):
@qinqon Thank you for all your work here, the tests help a lot as well. 

LGTM

Regarding a release of afterburn it might be a few weeks, as we just started our sprint, and have already planed the next 3 weeks. But it will be in our backlog to pick up.

jlebon reviewed (commented):
There's a lot there and I didn't go through it all, though one thing did catch my eye.
